crypto/defaults.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 1995-2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include "internal/thread_once.h"
+#include "internal/cryptlib.h"
+#include "internal/e_os.h"
+
+#if defined(_WIN32)
+
+/**
+ * @brief The directory where OpenSSL is installed.
+ */
+static char openssldir[MAX_PATH];
+
+/**
+ * @brief The directory where OpenSSL engines are located.
+ */
+
+static char enginesdir[MAX_PATH];
+
+/**
+ * @brief The directory where OpenSSL modules are located.
+ */
+static char modulesdir[MAX_PATH];
+
+/**
+ * @brief Get the list of Windows registry directories.
+ *
+ * This function retrieves a list of Windows registry directories.
+ *
+ * @return A pointer to a char array containing the registry directories.
+ */
+static char *get_windows_regdirs(void)
+{
+    DWORD keysize;
+    DWORD ktype;
+    HKEY hkey;
+    LSTATUS ret;
+    LPCWCH tempstr = NULL;
+
+    if (strlen(openssldir))
+      return openssldir;
+
+    ret = RegOpenKeyEx(HKEY_LOCAL_MACHINE, TEXT("SOFTWARE\\Wow6432Node\\OpenSSL"), 0, KEY_QUERY_VALUE, &hkey);
+    if (ret != ERROR_SUCCESS)
+      goto out;
+
+    ret = RegQueryValueEx(hkey, TEXT("OPENSSLDIR"), NULL, &ktype, NULL, &keysize);
+    if (ret != ERROR_SUCCESS)
+      goto out;
+    if (ktype != REG_EXPAND_SZ)
+      goto out;
+
+    keysize++;
+    tempstr = OPENSSL_zalloc(keysize * sizeof(TCHAR));
+
+    if (tempstr == NULL)
+      goto out;
+
+    if (RegQueryValueEx(hkey, TEXT("OPENSSLDIR"),
+                         NULL, &ktype, tempstr, &keysize) != ERROR_SUCCESS)
+        goto out;
```

Comment:
So if OPENSSLDIR is not present then we give up and don't attempt to read MODULESDIR or ENGINESDIR even if they might be present. But if OPENSSLDIR is present and MODULESDIR is not, we will use the OPENSSLDIR value and ignore any ENGINESDIR. This seems inconsistent.